### ISSUE TITLE
Bug: [@typescript-eslint/require-await]  Incorrectly reporting error for async iterator

### ISSUE THREAD

**Issue by yohny** [typescript-eslint/typescript-eslint#11669]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=PTAEGMHsCdoU3AF1NArgGzqAZgS2rgHYDmAsAFACGAzgJ6Hg6oOK6SFMMCMAFAJSgA3hVAo4iVNA4B2ANwUAvhWXkQoIlFgJkaTDnxEyVOg05I2hAFSga9cAElEcaJUQxeA4eVG1ccdAAmoFzy3qC%2B-kEATKE%2BfoGgAMyhSuQUUITUyLhOALbUoAC8oADaXAA0oFGViQC6obam2Mzm7NaNDk4ubtBR-EIiODCgPBlZ6nmgkNgTcPmeg3GRs7mxoKmpFEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEI0QR1gEtpEBaAQwHdzCAXdKAewDNnIAacbSAAVoE8ADigDG0QoNqkU8QgDtaAehIFiZKjXoZIiaNEbRIXAL4hjQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// rule reports error for below async iterator
async function* asyncIterator1() {
  yield 1;
  yield 2;
  yield 3;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "require-await": "off",
    "@typescript-eslint/require-await": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

rule does not report error for async iterator code

### Actual Result

rule considers async iterator code as incorrect

### Additional Info

async iterator on MDN https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncIterator

**Comment by kirkwaiblinger** [comment]
Hi @yohny!

Why exactly should this not be reported? See the note at the top of the [docs](https://typescript-eslint.io/rules/require-await):

> yield expressions in [async generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function*) behave differently from sync generator functions (they unwrap promises), so the base rule never checks async generator functions. On the other hand, our rule uses type information and can detect async generator functions that both never use await and always yield non-promise values.

**Comment by yohny** [comment]
hello,
the rule is correctly not reporting error when async generator is yielding promises
```ts
async function* asyncIterator1() {
  yield Promise.resolve(1);
  yield Promise.resolve(2);
  yield Promise.resolve(3);
}
```
but my understnding of async generator is, that it is also valid to yeild values directly (not promises), like in the code sample in the original issue. In such case the rule reports an error.
Async generator yielding values directly is also used as example at MDN https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function*#declaring_an_async_generator_function
but maybe this is misunderstanding on my side.

**Comment by kirkwaiblinger** [comment]
@yohny It is valid to yield values directly, but usually the reason for using an async generator would be because something async needs to happen in order to _get_ those values. For example, a real world use case I have encountered is

```ts
async function* getPaginatedApi() {
    let result = await fetchFromApi();
    yield result.data;
    while (result.hasMore) {
        result = await fetchFromApi(result.nextPageToken);
        yield result.data;
    }
}
```

So, in this example, one is actually _using_ the async capabilities of the generator with the `await` statements. 

If it didn't have the `await`s, and just yielded values, one might ask why it wasn't simply a sync generator.

**Comment by yohny** [comment]
ok, this is acceptable then,  I'm closing this issue

**Comment by kirkwaiblinger** [comment]
FWIW, the way I interpret the reports generated by `require-await` isn't "this code is invalid", it's more of "please double-check that this is intentional/necessary"
